### PR TITLE
PemToDer Bug Fixes

### PR DESCRIPTION
**Pull request by cconlon**
These four bug fixes are associated with Zendesk Ticket 2075.

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -3456,10 +3456,13 @@ int PemToDer(const unsigned char* buff, long longSz, int type,
 
     headerEnd += XSTRLEN(header);
 
+    if ((int)(headerEnd - (char*)buff) > sz - 2)
```
Comment (toddouska) [inline src/ssl.c]: What if sz is 1.  Then sz - 2 will not work.

File: src/ssl.c [hunk 2]
```diff
@@ -3456,10 +3456,13 @@ int PemToDer(const unsigned char* buff, long longSz, int type,
...
+        return SSL_BAD_FILE;
+
     /* eat end of line */
-    if (headerEnd[0] == '\n')
+    if (headerEnd && headerEnd[0] == '\n')
         headerEnd++;
-    else if (headerEnd[1] == '\n')
+    else if ((headerEnd + 1) && headerEnd[1] == '\n')
```
Comment (toddouska) [inline src/ssl.c]: headerEnd + 1 check is redundant.

File: src/ssl.c [hunk 3]
```diff
@@ -3534,7 +3534,7 @@ int PemToDer(const unsigned char* buff, long longSz, int type,
         /* eat end of line */
         if (consumedEnd[0] == '\n')
             consumedEnd++;
-        else if (consumedEnd[1] == '\n')
+        else if ((consumedEnd + 2 <= bufferEnd) && consumedEnd[1] == '\n')
```
Comment (toddouska) [inline src/ssl.c]: consumedEnd + 1 < bufferEnd is simpler here.

File: src/ssl.c [hunk 4]
```diff
@@ -3456,10 +3456,13 @@ int PemToDer(const unsigned char* buff, long longSz, int type,
 
     headerEnd += XSTRLEN(header);
 
+    if ((int)(headerEnd - (char*)buff) > sz - 2)
+        return SSL_BAD_FILE;
+
     /* eat end of line */
-    if (headerEnd[0] == '\n')
+    if (headerEnd && headerEnd[0] == '\n')
```
Comment (toddouska) [inline src/ssl.c]: If headerEnd is redundant here.  There is a !headerEnd check a few lines above making sure the pointer isn't NULL.  Adding to that pointer can't make it null.

File: src/ssl.c (hunk 1 above)
Comment (toddouska) [inline src/ssl.c]: headerEnd < (bufferEnd + 1)

Is a safer check.

**Comment by cconlon** [comment]
Thanks for reviewing, will make changes and update.

File: src/ssl.c (hunk 1 above)
Comment (cyassl) [inline src/ssl.c]: Make that a > , checking for overrun.

Comment (cconlon) [inline src/ssl.c]: Looks like this should be ">=", since reading from bufferEnd is out of bounds:

(headerEnd + 1) >= bufferEnd

**Comment by toddouska** [comment]
Thanks Chris.  Did you test the latest changes on the reported issues?

**Comment by cconlon** [comment]
Yes, the latest changes have been tested against the reported issues.  I can confirm they fix all submitted reproducible test cases.

**Pull request merged**